src/libmeasurement_kit/engine.cpp

```diff
@@ -0,0 +1,480 @@
+#include <mutex>
+#include <set>
+#include <string>
+#include <sstream>
+#include <thread>
+#include <tuple>
+#include <type_traits>
+
+#include <measurement_kit/common/error.hpp>
+#include <measurement_kit/common/logger.hpp>
+#include <measurement_kit/common/nlohmann/json.hpp>
+#include <measurement_kit/common/reactor.hpp>
+#include <measurement_kit/common/shared_ptr.hpp>
+
+#include "src/libmeasurement_kit/nettests/runnable.hpp"
+
+namespace mk {
+namespace engine {
+
+// # Multi-thread stuff
+//
+// Comes first because it needs more careful handling.
+
+class Semaphore {
+  public:
+    Semaphore() = default;
+
+    void wait() {
+        std::unique_lock<std::mutex> lock{mutex_};
+        cond_.wait(lock, [this]() { return !active_; });
+        active_ = true;
+    }
+
+    void signal() {
+        {
+            std::unique_lock<std::mutex> _{mutex_};
+            active_ = false;
+        }
+        cond_.notify_one();
+    }
+
+    ~Semaphore() = default;
+
+  private:
+    bool active_ = false;
+    std::condition_variable cond_;
+    std::mutex mutex_;
+};
+
+class TaskImpl {
+  public:
+    std::condition_variable cond;
+    std::deque<nlohmann::json> deque;
+    std::atomic_bool interrupted{false};
+    std::mutex mutex;
+    SharedPtr<Reactor> reactor = Reactor::make();
+    std::atomic_bool running{false};
+    std::thread thread;
+};
+
+static void task_run(TaskImpl *pimpl, nlohmann::json &settings);
+static bool is_event_valid(const std::string &);
+
+static void emit(TaskImpl *pimpl, nlohmann::json &&event) {
+
+    // In debug mode, make sure that we're emitting an event that we know
+    assert(event.is_object());
+    assert(event.count("type") != 0);
+    assert(event.at("type").is_string());
+    assert(is_event_valid(event.at("type").get<std::string>()));
+
+    // Actually emit the event.
+    {
+        std::unique_lock<std::mutex> _{pimpl->mutex};
+        pimpl->deque.push_back(std::move(event));
+    }
+    pimpl->cond.notify_all(); // more efficient if unlocked
+}
+
+Task::Task(nlohmann::json &&settings) {
+    pimpl_ = std::make_unique<TaskImpl>();
+    std::promise<void> barrier;
+    std::future<void> started = barrier.get_future();
+    pimpl_->thread = std::thread([this, &barrier,
+                                         settings = std::move(
+                                                 settings)]() mutable {
+        pimpl_->running = true;
+        barrier.set_value();
+        static Semaphore semaphore;
+        semaphore.wait(); // block until a previous task has finished running
+        task_run(pimpl_.get(), settings);
+        pimpl_->running = false;
+        pimpl_->cond.notify_all(); // tell the readers we're done
+        semaphore.signal();        // allow another task to run
+    });
+    started.wait(); // guarantee Task() completes when the thread is running
+}
+
+bool Task::is_running() const {
+    return pimpl_->running; // is an atomic var
+}
+
+void Task::interrupt() {
```

Comment:
From looking at the code it seems like the `interrupt` method is only something that is callable before the test starts, as such it's doesn't seem possible to interrupt a test mid-way. Or am I missing something?

Reply:
When the test is running, it's running on `pimpl_->reactor`. Calling `pimpl_->reactor->stop()` should cause the reactor to exit and so `run_with_initial_event()` should return and the code continue from there.

Reply:
Ah ok I understand now. I guess I was confused by there being some `begin` and `end` methods being called later, but I assume that is something that tis managed by the reactor.

Reply:
Yes, correct! I am planning to remove these `begin` and `end` methods as part of future refactoring as they're named most like this for historical reasons.